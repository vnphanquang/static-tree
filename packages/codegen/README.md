<div align="center">

# `static-tree-codegen`

</div>

This package is part of the [`static-tree` monorepo](https://github.com/vnphanquang/static-tre)

---

## Table of Contents
