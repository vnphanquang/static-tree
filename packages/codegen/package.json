{
  "name": "static-tree-codegen",
  "version": "0.0.0",
  "description": "Code generation for static-tree",
  "main": "lib/index.js",
  "module": "lib/index.esm.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "build": "rollup -c",
    "lint": "eslint **/*.{ts,js}",
    "format": "prettier --check \"./**/*.{ts,js,json}\"",
    "test": "vitest --ui --coverage",
    "ci:test": "vitest run --coverage",
    "api:extract": "api-extractor run --local --verbose",
    "api:document": "api-documenter markdown -i api/temp -o api/docs",
    "api": "run-s api:extract api:document"
  },
  "keywords": [
    "typescript",
    "static-tree",
    "codegen"
  ],
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vnphanquang/static-tree.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "bugs": {
    "url": "https://github.com/vnphanquang/static-tree/issues"
  },
  "homepage": "https://github.com/vnphanquang/static-tree/tree/main/packages/codegen",
  "author": "Quang Phan",
  "license": "MIT",
  "dependencies": {
    "static-tree": "next",
    "typescript": "^4.7.4"
  },
  "devDependencies": {
    "@microsoft/api-documenter": "^7.19.5",
    "@microsoft/api-extractor": "^7.29.3",
    "@rollup/plugin-commonjs": "^22.0.2",
    "@rollup/plugin-node-resolve": "^13.3.0",
    "@types/node": "^18.7.3",
    "@vitest/coverage-istanbul": "^0.22.1",
    "@vitest/ui": "^0.22.0",
    "@vnphanquang/eslint-config": "^1.1.0",
    "eslint": "^8.22.0",
    "npm-run-all": "^4.1.5",
    "prettier": "^2.7.1",
    "rollup": "^2.78.1",
    "rollup-plugin-filesize": "^9.1.2",
    "rollup-plugin-typescript2": "^0.32.1",
    "ts-node": "^10.9.1",
    "vite": "^3.0.8",
    "vitest": "^0.22.0"
  }
}
