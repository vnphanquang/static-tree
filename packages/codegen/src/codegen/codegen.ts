import type { SerializedTNode } from 'static-tree';
import {
  createPrinter,
  createSourceFile,
  factory,
  ListFormat,
  NewLineKind,
  NodeFlags,
  ScriptKind,
  ScriptTarget,
  SyntaxKind,
} from 'typescript';

const PACKAGE = 'static-tree';
const BUILDER_HELPER = 'tBuild';

export function generateImportStatement() {
  return factory.createImportDeclaration(
    undefined,
    undefined,
    factory.createImportClause(
      false,
      undefined,
      factory.createNamedImports([
        factory.createImportSpecifier(false, undefined, factory.createIdentifier(BUILDER_HELPER)),
      ]),
    ),
    factory.createStringLiteral(PACKAGE),
    undefined,
  );
}

export function generateBuildInput(serializedNode: SerializedTNode) {
  const { children, key } = serializedNode;

  if (children.length === 0) {
    return factory.createStringLiteral(key);
  } else {
    const keyPropAssignment = factory.createPropertyAssignment(
      factory.createIdentifier('key'),
      factory.createStringLiteral(key),
    );
    const [firstChild, ...otherChildren] = children;
    let builderCallExpression = factory.createCallExpression(
      factory.createPropertyAccessExpression(
        factory.createIdentifier('builder'),
        factory.createIdentifier('addChild'),
      ),
      undefined,
      [generateBuildInput(firstChild)],
    );
    for (const child of otherChildren) {
      builderCallExpression = factory.createCallExpression(
        factory.createPropertyAccessExpression(
          builderCallExpression,
          factory.createIdentifier('addChild'),
        ),
        undefined,
        [generateBuildInput(child)],
      );
    }

    const buildPropAssignment = factory.createPropertyAssignment(
      factory.createIdentifier('build'),
      factory.createArrowFunction(
        undefined,
        undefined,
        [
          factory.createParameterDeclaration(
            undefined,
            undefined,
            undefined,
            factory.createIdentifier('builder'),
            undefined,
            undefined,
            undefined,
          ),
        ],
        undefined,
        factory.createToken(SyntaxKind.EqualsGreaterThanToken),
        builderCallExpression,
      ),
    );

    return factory.createObjectLiteralExpression([keyPropAssignment, buildPropAssignment], true);
  }
}

export function generateTBuild(serializedNode: SerializedTNode) {
  const buildInput = generateBuildInput(serializedNode);
  return factory.createCallExpression(factory.createIdentifier(BUILDER_HELPER), undefined, [
    buildInput,
  ]);
}

export function generate(serializedNode: SerializedTNode) {
  const sourceFile = createSourceFile('', '', ScriptTarget.Latest, false, ScriptKind.TS);
  const printer = createPrinter({
    newLine: NewLineKind.LineFeed,
  });

  const nodeAssignment = factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createObjectBindingPattern([
            factory.createBindingElement(
              undefined,
              undefined,
              factory.createIdentifier('node'),
              undefined,
            ),
          ]),
          undefined,
          undefined,
          generateTBuild(serializedNode),
        ),
      ],
      NodeFlags.Const,
    ),
  );

  const printed = printer.printList(
    ListFormat.MultiLine,
    factory.createNodeArray([generateImportStatement(), nodeAssignment]),
    sourceFile,
  );

  return printed;
}
