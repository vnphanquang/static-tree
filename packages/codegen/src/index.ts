// Copyright (c) Quang Phan. All rights reserved. Licensed under the MIT license.

/**
 * Code generation for package {@link https://github.com/vnphanquang/static-tree/tree/main/packages/static-tree static-tree}
 *
 * @packageDocumentation
 */

export * from './codegen';
