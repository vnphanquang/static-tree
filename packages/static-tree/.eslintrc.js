module.exports = {
  root: true,
  extends: '@vnphanquang/eslint-config',
  parserOptions: {
    project: 'tsconfig.json',
    tsconfigRootDir: __dirname,
    sourceType: 'module',
  },
  env: {
    node: true,
    jest: true,
  },
  ignorePatterns: [],
  rules: {},
};
