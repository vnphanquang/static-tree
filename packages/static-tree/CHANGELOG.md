# static-tree

## 1.3.1

### Patch Changes

- [`e758663`](https://github.com/vnphanquang/static-tree/commit/e758663b9c832f2e5246628db1bdb47fd7819933) Thanks [@vnphanquang](https://github.com/vnphanquang)! - fix typescript build output (rollup)

## 1.3.0

### Minor Changes

- [`67bb6c2`](https://github.com/vnphanquang/static-tree/commit/67bb6c26704c26ee0b26d68e688e5b51c24682b8) Thanks [@vnphanquang](https://github.com/vnphanquang)! - Added support for runtime dynamic path segment resolution

## 1.2.0

### Minor Changes

- [#5](https://github.com/vnphanquang/static-tree/pull/5) [`8544a4c`](https://github.com/vnphanquang/static-tree/commit/8544a4c4b468262109ecb4139f4a720c6b23b2a1) Thanks [@vnphanquang](https://github.com/vnphanquang)! - [breaking] separate key as mandatory own function argument for better type support

### Patch Changes

- [#5](https://github.com/vnphanquang/static-tree/pull/5) [`8544a4c`](https://github.com/vnphanquang/static-tree/commit/8544a4c4b468262109ecb4139f4a720c6b23b2a1) Thanks [@vnphanquang](https://github.com/vnphanquang)! - should pass `pathResolver` on nested children

## 1.1.1

### Patch Changes

- [`8460174`](https://github.com/vnphanquang/static-tree/commit/8460174dccfb73751b6e1aad80ee4be74a892e8f) Thanks [@vnphanquang](https://github.com/vnphanquang)! - fix versioning

## 1.0.1

### Patch Changes

- [`d1b5241`](https://github.com/vnphanquang/static-tree/commit/d1b5241758ccd57c96d76187f2bdbaad286fc8b4) Thanks [@vnphanquang](https://github.com/vnphanquang)! - change homepage link for static-tree package

## 1.1.0

### Minor Changes

- [`a6b1fb5`](https://github.com/vnphanquang/static-tree/commit/a6b1fb51aca76080fe61927b40e6459e6e91a87d) Thanks [@vnphanquang](https://github.com/vnphanquang)! - rename builder helper to `tBuild` and allow multiple input types

### Patch Changes

- [`2dba2cf`](https://github.com/vnphanquang/static-tree/commit/2dba2cfa64debf358ee287415bf637c10ee572b5) Thanks [@vnphanquang](https://github.com/vnphanquang)! - add README docs

* [`1387577`](https://github.com/vnphanquang/static-tree/commit/1387577cf9b575860418bf49f244c71dd029b22f) Thanks [@vnphanquang](https://github.com/vnphanquang)! - (TNode): add option to reverse built path

- [`dc1210f`](https://github.com/vnphanquang/static-tree/commit/dc1210fa3719e1869ad383b54db055edbd6a3421) Thanks [@vnphanquang](https://github.com/vnphanquang)! - add `pathResolver` option to `TNodeInit`

* [`3c7321e`](https://github.com/vnphanquang/static-tree/commit/3c7321e9e6fec2fc183a8609b1a014889669bece) Thanks [@vnphanquang](https://github.com/vnphanquang)! - docs: bootstrap tsdoc and api extractor scripts

- [`a8a9169`](https://github.com/vnphanquang/static-tree/commit/a8a9169a7e0c5a60226343e0307961a231e93834) Thanks [@vnphanquang](https://github.com/vnphanquang)! - enable `till` option in TNode get path method

## 1.0.0

### Major Changes

- [`522ba90`](https://github.com/vnphanquang/static-tree/commit/522ba90f17803778d5c6ca2454135a2069cc6650) Thanks [@vnphanquang](https://github.com/vnphanquang)! - initial implementation of static tree builder
