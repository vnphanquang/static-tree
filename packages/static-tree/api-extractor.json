/**
 * Config file for API Extractor. For more info, please visit: https://api-extractor.com/pages/configs/api-extractor_json/
 */
{
  "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
  "projectFolder": ".",
  "mainEntryPointFilePath": "<projectFolder>/lib/index.d.ts",
  "bundledPackages": [],
  "compiler": {
    "tsconfigFilePath": "<projectFolder>/tsconfig.json"
  },
  "apiReport": {
    "enabled": true,
    "reportFileName": "<unscopedPackageName>.api.md",
    "reportFolder": "<projectFolder>/api/",
    "reportTempFolder": "<projectFolder>/api/temp/"
  },
  "docModel": {
    "enabled": true,
    "apiJsonFilePath": "<projectFolder>/api/temp/<unscopedPackageName>.api.json"
  },
  "dtsRollup": {
    "enabled": false
  },
  "tsdocMetadata": {
    "enabled": true
  },
  "newlineKind": "lf",
  "messages": {
    "compilerMessageReporting": {
      "default": {
        "logLevel": "warning"
        // "addToApiReportFile": false
      }
    },
    "extractorMessageReporting": {
      "default": {
        "logLevel": "warning"
        // "addToApiReportFile": false
      }
    },
    "tsdocMessageReporting": {
      "default": {
        "logLevel": "warning"
        // "addToApiReportFile": false
      }
    }
  }
}
