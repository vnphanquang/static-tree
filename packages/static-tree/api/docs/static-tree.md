<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [static-tree](./static-tree.md)

## static-tree package

Type-safe static tree builder

## Classes

|  Class | Description |
|  --- | --- |
|  [ExtendedTNodeBuilder](./static-tree.extendedtnodebuilder.md) | Builder for [ExtendedTNode](./static-tree.extendedtnode.md) |
|  [TNode](./static-tree.tnode.md) | Implementation for each node of the static tree |

## Functions

|  Function | Description |
|  --- | --- |
|  [tBuild(builder)](./static-tree.tbuild.md) | Build helper that uses [ExtendedTNodeBuilder](./static-tree.extendedtnodebuilder.md) internally. |
|  [tBuild(key, config)](./static-tree.tbuild_1.md) |  |

## Interfaces

|  Interface | Description |
|  --- | --- |
|  [ExtendedTNodeBuildConfig](./static-tree.extendedtnodebuildconfig.md) | Instruction for building nested [ExtendedTNode](./static-tree.extendedtnode.md)<!-- -->. |
|  [MinimalSerializedTNode](./static-tree.minimalserializedtnode.md) | Interface of a minimally serialized node |
|  [TBuildOutput](./static-tree.tbuildoutput.md) | Output of  |
|  [TNodeGetPathParams](./static-tree.tnodegetpathparams.md) | Instruction on how to construct path to / from current node |
|  [TNodeInit](./static-tree.tnodeinit.md) | Additional instruction during [TNode](./static-tree.tnode.md) construction |
|  [TNodeInternalApi](./static-tree.tnodeinternalapi.md) | The proxy api intended for internal or advanced use Refer to [TNode](./static-tree.tnode.md) for documentation of each method. |
|  [TNodePublicApi](./static-tree.tnodepublicapi.md) | The proxy api intended for public use. Refer to [TNode](./static-tree.tnode.md) for documentation of each method. |
|  [TNodeSerializeOptions](./static-tree.tnodeserializeoptions.md) | Instruction on how to serialize the current node |
|  [VerboseSerializedTNode](./static-tree.verboseserializedtnode.md) | Interface of a verbosely serialized node |

## Type Aliases

|  Type Alias | Description |
|  --- | --- |
|  [ExtendedTNode](./static-tree.extendedtnode.md) | <p>[TNode](./static-tree.tnode.md) extended with children inline as properties, intended to provide idiomatic dot notation parent-to-child node access, so that you can do:</p>
```typescript
root.child.grandChild.$.path(); // -\> `root/child/grandChild`
```
<p>instead of:</p>
```typescript
root.$.children().find(c => c.key === 'child')?.$.children().find(c => c.key === 'grandChild')?.$.path();
```
 |
|  [ExtendedTNodeBuildCallback](./static-tree.extendedtnodebuildcallback.md) | Build callback for adding data &amp; nested children |
|  [SerializedTNode](./static-tree.serializedtnode.md) | Possible types of serialized node, depending on how [TNodeSerializeOptions](./static-tree.tnodeserializeoptions.md) is setup |
|  [TNodeData](./static-tree.tnodedata.md) | Inner data of node. See  for idiomatic examples on adding data to node. See [TNode](./static-tree.tnode.md) for more information on how this data is stored. |

