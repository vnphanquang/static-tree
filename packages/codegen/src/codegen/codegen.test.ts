import { tBuild } from 'static-tree';
import { expect, test } from 'vitest';

import { generate } from './codegen';

const { node } = tBuild({
  key: 'api',
  pathResolver: () => 'https://api.domain.example',
  build: (builder) =>
    builder
      .addData({
        number: 1,
        string: 'string',
        boolean: true,
        object: {
          nested: 'nested',
        },
        array: [
          {
            id: 'test',
          },
        ],
      })
      .addChild({
        key: 'auth',
        build: (builder) =>
          builder
            .addChild('logout')
            .addChild({
              key: 'oauth',
              build: (builder) =>
                builder
                  .addChild('google')
                  .addChild('discord')
                  .addChild('facebook')
                  .addChild('github'),
            })
            .addChild('login'),
      })
      .addChild('healthcheck'),
});

test('generate', () => {
  expect(generate(node.$.serialize())).toMatchSnapshot();
});
