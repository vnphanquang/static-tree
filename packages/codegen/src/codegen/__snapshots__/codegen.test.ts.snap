// Vitest Snapshot v1

exports[`generate 1`] = `
"import { tBuild } from \\"static-tree\\";
const { node } = tBuild({
    key: \\"api\\",
    build: builder => builder.addChild({
        key: \\"auth\\",
        build: builder => builder.addChild(\\"logout\\").addChild({
            key: \\"oauth\\",
            build: builder => builder.addChild(\\"google\\").addChild(\\"discord\\").addChild(\\"facebook\\").addChild(\\"github\\")
        }).addChild(\\"login\\")
    }).addChild(\\"healthcheck\\")
});
"
`;
